### PR TITLE
[AF-2373] Squash Commits on Change request

### PR DESCRIPTION
**Pull request by gabriel376**
[AF-2373 Squash Commits on Change request](https://issues.redhat.com/browse/AF-2373)

AF-2373

### CODE REVIEW

**Comment by barboras7** [comment]
jenkins execute full downstream build

File: uberfire-structure/uberfire-structure-api/src/main/java/org/guvnor/structure/repositories/changerequest/portable/ChangeRequestCommit.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+import org.jboss.errai.common.client.api.annotations.Portable;
+
+@Portable
+public class ChangeRequestCommit {
+
+    private String id;
+    private String message;
+
+    public ChangeRequestCommit() {
```
Comment (caponetto) [inline uberfire-structure/uberfire-structure-api/src/main/java/org/guvnor/structure/repositories/changerequest/portable/ChangeRequestCommit.java]: Consider removing this empty constructor.

File: uberfire-structure/uberfire-structure-api/src/main/java/org/guvnor/structure/repositories/changerequest/portable/ChangeRequestCommit.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    }
+
+    public ChangeRequestCommit(final String id,
```
Comment (caponetto) [inline uberfire-structure/uberfire-structure-api/src/main/java/org/guvnor/structure/repositories/changerequest/portable/ChangeRequestCommit.java]: Consider annotating the parameters with @MapsTo.

File: uberfire-structure/uberfire-structure-backend/src/main/java/org/guvnor/structure/backend/repositories/ChangeRequestServiceImpl.java [hunk 3]
```diff
@@ -1333,4 +1334,61 @@ JGitPathImpl createJGitPathImpl(final JGitFileSystem fs,
...
+                                                                 repositoryAlias,
+                                                                 false,
+                                                                 changeRequestId);
+
+        final Git git = getGitFromBranch(repository,
+                                         changeRequest.getSourceBranch());
+
+        final String startCommit = changeRequest.getStartCommitId();
+
+        git.squash(changeRequest.getSourceBranch(),
```
Comment (caponetto) [inline uberfire-structure/uberfire-structure-backend/src/main/java/org/guvnor/structure/backend/repositories/ChangeRequestServiceImpl.java]: Is it ok to squash the source branch? I mean, the whole history will be lost.

File: uberfire-structure/uberfire-structure-backend/src/main/java/org/guvnor/structure/backend/repositories/ChangeRequestServiceImpl.java [hunk 4]
```diff
@@ -1333,4 +1334,61 @@ JGitPathImpl createJGitPathImpl(final JGitFileSystem fs,
...
+                   startCommit,
+                   commitMessage);
+
+        return mergeChangeRequest(spaceName,
```
Comment (caponetto) [inline uberfire-structure/uberfire-structure-backend/src/main/java/org/guvnor/structure/backend/repositories/ChangeRequestServiceImpl.java]: Question: what happens after the squash+merge if there is only one commit?

File: uberfire-structure/uberfire-structure-backend/src/main/java/org/guvnor/structure/backend/repositories/ChangeRequestServiceImpl.java [hunk 5]
```diff
@@ -1333,4 +1334,61 @@ JGitPathImpl createJGitPathImpl(final JGitFileSystem fs,
                                    objectId,
                                    isRealPath);
     }
-}
\ No newline at end of file
+
+    @Override
+    public List<ChangeRequestCommit> getCommits(final String spaceName,
```
Comment (caponetto) [inline uberfire-structure/uberfire-structure-backend/src/main/java/org/guvnor/structure/backend/repositories/ChangeRequestServiceImpl.java]: Consider creating at least one unit test for this public method.

File: uberfire-structure/uberfire-structure-backend/src/main/java/org/guvnor/structure/backend/repositories/ChangeRequestServiceImpl.java (hunk 4 above)
Comment (gabriel376) [reply uberfire-structure/uberfire-structure-backend/src/main/java/org/guvnor/structure/backend/repositories/ChangeRequestServiceImpl.java]: works fine, I could not see any edge case here

File: uberfire-structure/uberfire-structure-backend/src/main/java/org/guvnor/structure/backend/repositories/ChangeRequestServiceImpl.java (hunk 3 above)
Comment (gabriel376) [reply uberfire-structure/uberfire-structure-backend/src/main/java/org/guvnor/structure/backend/repositories/ChangeRequestServiceImpl.java]: I don't know, there is no spec

Comment (caponetto) [reply uberfire-structure/uberfire-structure-backend/src/main/java/org/guvnor/structure/backend/repositories/ChangeRequestServiceImpl.java]: Not sure if the version history is useful for users after the merge. If so, then we can merge first into the target branch and then squash the commits there (target branch).
@paulovmr ^

File: uberfire-structure/uberfire-structure-api/src/main/java/org/guvnor/structure/repositories/changerequest/portable/ChangeRequestCommit.java (hunk 1 above)
Comment (gabriel376) [reply uberfire-structure/uberfire-structure-api/src/main/java/org/guvnor/structure/repositories/changerequest/portable/ChangeRequestCommit.java]: Done

File: uberfire-structure/uberfire-structure-api/src/main/java/org/guvnor/structure/repositories/changerequest/portable/ChangeRequestCommit.java (hunk 2 above)
Comment (gabriel376) [reply uberfire-structure/uberfire-structure-api/src/main/java/org/guvnor/structure/repositories/changerequest/portable/ChangeRequestCommit.java]: Done

File: uberfire-structure/uberfire-structure-backend/src/main/java/org/guvnor/structure/backend/repositories/ChangeRequestServiceImpl.java (hunk 5 above)
Comment (gabriel376) [reply uberfire-structure/uberfire-structure-backend/src/main/java/org/guvnor/structure/backend/repositories/ChangeRequestServiceImpl.java]: Done

File: uberfire-structure/uberfire-structure-backend/src/main/java/org/guvnor/structure/backend/repositories/ChangeRequestServiceImpl.java (hunk 4 above)
Comment (Srihari1192) [reply uberfire-structure/uberfire-structure-backend/src/main/java/org/guvnor/structure/backend/repositories/ChangeRequestServiceImpl.java]: After merging one commit and trying to do squash , getting no response from page.Upon refreshing and navigating to Change request  getting error as "The source and target branches are in conflict, thus cannot be integrated."
But same squash commit message can be seen under version.


After 1st merge and squash error
After 1st merge and squash under version

Comment (gabriel376) [reply uberfire-structure/uberfire-structure-backend/src/main/java/org/guvnor/structure/backend/repositories/ChangeRequestServiceImpl.java]: Hi @Srihari1192 
Thank you for catching this up
I could not reproduce the scenario you are describing. Can you provide more details ?
Merge + Squash a single commit works fine for me

Comment (Srihari1192) [reply uberfire-structure/uberfire-structure-backend/src/main/java/org/guvnor/structure/backend/repositories/ChangeRequestServiceImpl.java]: Steps:
1.add asset data object.
2.Submit change request.
3.accept the change request with master(Note: not doing squash just merge).
4.Again modify asset and create change request.
5.Now try to squash the changes  with master and after click on 'Ok', not showing any response
6.After step 5-> we are able to see squash message under Version but change request is prompting still has "The source and target branches are in conflict, thus cannot be integrated.".
So two observations step 5 and step 6..Please refer screenshot for reference
Step1
Step2
Step3
Step4
Step5
Step6
Step6_1
Step 4_1

**Comment by Srihari1192** [comment]
jenkins execute full downstream build

**Comment by Srihari1192** [comment]
@all...When we do  squashing the commits and click on 'Ok' , change request is accepting without click on merge.Is this behavior correct?
As per acceptance criteria the change request should accept after click on merge and in squash just user should be able to modify  the commit message.

File: uberfire-structure/uberfire-structure-backend/src/main/java/org/guvnor/structure/backend/repositories/ChangeRequestServiceImpl.java (hunk 3 above)
Comment (paulovmr) [reply uberfire-structure/uberfire-structure-backend/src/main/java/org/guvnor/structure/backend/repositories/ChangeRequestServiceImpl.java]: @gabriel376 @caponetto I think it's better to keep the history, so the squash could be done in the target branch.

Comment (gabriel376) [reply uberfire-structure/uberfire-structure-backend/src/main/java/org/guvnor/structure/backend/repositories/ChangeRequestServiceImpl.java]: Ok!

Comment (gabriel376) [reply uberfire-structure/uberfire-structure-backend/src/main/java/org/guvnor/structure/backend/repositories/ChangeRequestServiceImpl.java]: Done

File: uberfire-structure/uberfire-structure-backend/src/main/java/org/guvnor/structure/backend/repositories/ChangeRequestServiceImpl.java (hunk 4 above)
Comment (gabriel376) [reply uberfire-structure/uberfire-structure-backend/src/main/java/org/guvnor/structure/backend/repositories/ChangeRequestServiceImpl.java]: I submit a new commit to solve this issue
Let me know if this works

**Comment by paulovmr** [comment]
Hi @Srihari1192 , this behavior is correct, I've updated the acceptance criteria to reflect that. Thanks!

**Comment by paulovmr** [comment]
jenkins execute full downstream build

**Comment by Srihari1192** [comment]
jenkins execute full downstream build

**Comment by barboras7** [comment]
Hi guys, 

@Srihari1192 is taking care of this PR. I only wanted to share one concern. The two new buttons - Squash and Merge. Clicking "Squash" does the git squash on commits and then merges them. As a user, I would not expect button "Squash" to actually merge the PR. Maybe some tooltip would be nice or some alert to the user. Just my opinion on this :)

**Comment by paulovmr** [comment]
@barboras7 Ok I see, I think we can change the button label to "Squash and merge" then.

@gabriel376 @caponetto @ederign ^

**Comment by caponetto** [comment]
@barboras7 very good point
+1 for @paulovmr suggestion

**Comment by ederign** [comment]
+1

**Comment by gabriel376** [comment]
Hi @barboras7 
Thank you for the suggestion!
I submit a new commit on kie-wb-common to address the button label

File: uberfire-structure/uberfire-structure-backend/src/main/java/org/guvnor/structure/backend/repositories/ChangeRequestServiceImpl.java (hunk 4 above)
Comment (Srihari1192) [reply uberfire-structure/uberfire-structure-backend/src/main/java/org/guvnor/structure/backend/repositories/ChangeRequestServiceImpl.java]: @gabriel376 Now during creation of change request getting error as "The source and target branches are in conflict!" but same we can able to perform successfully, if we do before merge and try to do multiple merge.
Steps:
1.Add asset with created branch.
2.Create change request and Squash the commit.
3.Again modify asset and try to create change request.
4.getting error as "The source and target branches are in conflict!" .

Note:Same steps is getting successful, if we perform only merge actions.

**Comment by Srihari1192** [comment]
@all After first squash, we are not able to create change request, getting error as "The source and target branches are in conflict!". Looks like a bug.
But same we are able to create multiple change request ,if we perform only merge action.

**Comment by Srihari1192** [comment]
jenkins execute full downstream build

**Comment by Srihari1192** [comment]
jenkins execute full downstream build

**Comment by Srihari1192** [comment]
jenkins execute full downstream build

**Comment by Srihari1192** [comment]
@all Can someone confirm this observation..
while merging, we are able to create and merge from branch into master multiple times.. but when squashing, we are able to squash from branch to master only once, the second time getting a conflict..
Please find the recorded video link: https://drive.google.com/open?id=16b9tdgA3at-ulPaSZA_vfo1_87iQ6wsU

**Comment by caponetto** [comment]
Hi @Srihari1192,
Both Merge and Squash operations must have the same behavior.

**Comment by Srihari1192** [comment]
@gabriel376  Could you please look into the above behavior.

**Comment by gabriel376** [comment]
@Srihari1192,

This is expected behavior, it's not possible to squash multiple times the same branch into master. This happens because the squash operation rewrites git history, which creates a conflict between the branch and the master. Merge operation does not rewrite history and thus can be done multiple times into master.
You can easily reproduce this scenario through command line or though the github interface.

**Comment by caponetto** [comment]
If there is no way around this behavior, then I think we should somehow warn the user. In my understanding, the source branch becomes kind of useless after the squash, since we don't provide a conflict solver mechanism, thus could never be integrated again.

**Comment by Srihari1192** [comment]
jenkins execute full downstream build

**Comment by Srihari1192** [comment]
@gabriel376  Validation is not taken care for squash and merge text box, in case of if user enters null.
Squash_Merge
Squash_Merge_version

**Comment by gabriel376** [comment]
Hi @Srihari1192,
Good catch!
This is now fixed on kie-wb-common repository.

**Comment by adrielparedes** [comment]
jenkins execute full downstream build

**Comment by Srihari1192** [comment]
Hi @gabriel376  
If user enters null or empty and click on 'OK' , error message is not prompting, its closing 'Squash and Merge' popup.

**Comment by paulovmr** [comment]
Commits in this PR were incorporated [here](https://github.com/kiegroup/appformer/pull/884).

**Pull request closed without merging**